client/app/controllers/collection/add.js

```diff
@@ -133,3 +143,100 @@ export default Ember.Controller.extend({
 
     }
 });
+
+
+
+// Engine helper functions.
+// ////////////////////////////////////////////////////////////////////////////////////////////////
+
+
+function condition_dispatcher(condition) {
+
+    const parameters = this.get('parameters');
+
+    // Check if its a regular condition
+    if (condition.parameter !== undefined) {
+        // actualy check the condition is met;
+        // the parameter has to have the given state.
+        if (parameters[condition.parameter] === undefined) {
+            parameters[condition.parameter] = {};
+        }
+        if (parameters[condition.parameter].state === undefined) {
+            parameters[condition.parameter].state = [];
+        }
+        const parameter_state = parameters[condition.parameter].state
+        // check that the state exists for this item
+        return parameter_state.some((state_item) => state_item === condition.state)
+    }
+
+    // Check if its an 'all' composite condition
+    if (condition.all !== undefined &&
+        condition.all.constructor === Array
+    ) {
+        // if any conditions fail, the whole check fails.
+        return check_all.call(this, condition.all);
+    }
+
+    // Check if its an 'any' composite condition
+    if (condition.any !== undefined &&
+        condition.any.constructor === Array
+    ) {
+        // if any conditions are met, the whole check passes.
+        return check_any.call(this, condition.any);
+    }
+
+    // Check if its a 'none' composite condition
+    if (condition.none !== undefined &&
+        condition.none.constructor ===Array
+    ) {
+        // if any conditions are met, the whole check fails.
+        return !check_any.call(this, condition.none);
+    }
+
+    return false;
+
+}
+
+function check_all(conditions) {
+
+    const parameters = this.get('parameters');
+
+    if (typeof conditions !== 'object') return false;
+    if (conditions.constructor !== Array) return false;
+    // if any conditions fail, the whole check fails.
+    return !conditions.some(condition => !condition_dispatcher.call(this, condition));
+}
+
+function check_any(conditions) {
+    // If any conditions are met, the whole check passes.
+    return action.conditions.some(condition_dispatcher.bind(this))
+}
+
+
+function cons_arg_arr(action) {
```

Comment:
This is how the signature allows an obj to be used as the params for a fn. this also allows for hardcoded args as well as params to be described on the model. parameters are analagous to variables that are on the controller, args are analagous to hardcoded values that aren't dependent on the operation of the form.